Initiate client command

aarondufall opened the pull request.

aarondufall committed:
Initiate client command

sbellware commented on lib/file_transfer_component/commands/command.rb:
```diff
@@ -0,0 +1,32 @@
+        cls.class_exec do
+          include Messaging::StreamName
+          include FileTransferComponent::Messages::Commands
+          include FileTransferComponent::Messages::Events
```
I don't think you'll need access to the `Events` namespace here. The commands will likely only write commands.

sbellware reviewed (commented):
Aside from the inclusion of the Events namespace, and the generation of the file ID that we talked about on Slack, looks good.

aarondufall committed:
Add uri control

sbellware commented on lib/file_transfer_component/commands/initiate.rb:
```diff
@@ -5,16 +5,18 @@ class Initiate
 
-      def self.build(file_id, name, uri, reply_stream_name: nil)
+      def self.build(name, uri, reply_stream_name: nil)
+        file_id = Identifier::UUID::Random.get
```
Technically, the identifier should be a dependency. If it's not a dependency, you won't be able to _force_ the UUID to a specific value, which is useful for certain testing scenarios. That said, if you're not facing those testing scenarios, then using the constant directly, in a way that doesn't allow for substitution, should be fine. You could always change it later if you needed to. However, the design doesn't presently express the aspects of this class that are things that should be substitutable dependencies.

sbellware commented on lib/file_transfer_component/commands/initiate.rb:
```diff
@@ -5,16 +5,18 @@ class Initiate
+      def self.call(name, uri, reply_stream_name: nil)
+        instance = build(name, uri, reply_stream_name: reply_stream_name)
         instance.()
+        instance.file_id
```
This line should not be here. The class actuator should return whatever the instance actuator returns. So, if the class actuator should return the file ID, then the instance actuator should, as well, As it stands, this design has a significant inconsistency in the type signature between the class and instance actuators. That's nothing that the language can enforce. It's something that the programmer has to enforce based on first principals.

The class actuator should end at line 18, and then the instance actuator should be changed to reflect the natural behavior of the class.

Remember, the class actuator should only _delegate_ to the instance actuator. It should not (really _must not_) reinterpret the results of the instance actuator. The instance actuator is the definitive and authoritative truth. It's what the instance says that is the truth. The class methods are just there for convenience (especially in regards to construction).

sbellware commented on test/interactive/initiate.rb:
```diff
@@ -1,14 +1,12 @@
 require_relative '../client_test_init'
 
-file_id = ENV['FILE_ID']
```
This brings up some questions for me that might clarify the authoritative source of the file ID.

If you ever need to _replace_ a file (not that this is a good idea, but there might be some operations case where it's necessary; some emergency), the file ID may need to be passed in rather than generated.

Such a case should require that the present design be reconsidered. There might be two commands, rather than a single "initiate" command. You may need a "replace" command. In that case "initiate" may no longer be the right name for the current command, as a _replacement_ can be _initiated_ as well.

sbellware replied:
NOTE: I'm thinking more about a CLI used by operations staff rather than an interactive test. Reading this test made me thing about this.

aarondufall committed:
diagnostic sctipt

aarondufall committed:
Make uuid subsitutable

sbellware commented on lib/file_transfer_component/commands/initiate.rb:
```diff
@@ -2,12 +2,12 @@ module FileTransferComponent
+      attr_writer :file_id
 
-      initializer :file_id, :name, :uri
+      initializer :name, :uri
```
Seems like  file_id should be an initializer parameter. Otherwise, it can be left blank. ie: right now it's optional for this object to exist without a file ID.

sbellware commented on lib/file_transfer_component/commands/initiate.rb:
```diff
@@ -16,7 +16,16 @@ def self.build(name, uri, reply_stream_name: nil)
+      end
+
+      def file_id
+        @file_id ||= Identifier::UUID::Random.get
```
`Identifier::UUID::Random` should be a `dependency`.

This line would become:
`@file_id ||= identifier.get`

sbellware replied:
That said, it still needs to be decided wither the file ID is provided as an initializer argument.

sbellware commented on test/interactive/upload.rb:
```diff
@@ -0,0 +1,28 @@
+  handler.(event_data)
+end
+
+EventSource::Postgres::Read.("fileTransferComponent") do |event_data|
```
This block will not get executed. The above read is a blocking operation.

aarondufall committed:
Make indentify a dependency

sbellware commented on lib/file_transfer_component/commands/initiate.rb:
```diff
@@ -6,6 +6,8 @@ class Initiate
 
       initializer :name, :uri
 
+      dependency :identifier, Identifier::UUID::Random
```
Don't forget to assign an instance of the Udentifier::UUID::Random to this dependency attribute in the build method

sbellware commented on lib/file_transfer_component/commands/initiate.rb:
```diff
@@ -0,0 +1,43 @@
+
+      def call
+        stream_name = command_stream_name(file_id)
+        write.(command, stream_name, reply_stream_name: reply_stream_name)
```
Move above two lines to a method named `write` in the `Command` generalization, and then call `write` from the actuator.

aarondufall committed:
extract the write command logic

aarondufall committed:
Allow notes file

aarondufall committed:
Basic functionality

aarondufall committed:
update libs

aarondufall committed:
S3 Specs and controls

aarondufall committed:
Test not found with substitute

aarondufall committed:
add message and controls to client

aarondufall committed:
Add client gemspec

gem build

require all file except handlers

make aws-sdk a runtime dep

fix load order

add client loader

require all the things

temp client patch

new gemspec

yolo

yolo

yolo

yolo

yolo

yolo

yolo
